src/ejabberd_c2s.erl

```diff
@@ -1311,6 +1311,35 @@ handle_incoming_message(Info, StateName, StateData) ->
     ?ERROR_MSG("Unexpected info: ~p", [Info]),
     fsm_next_state(StateName, StateData).
 
+process_incoming_stanza_with_conflict_check(Name, From, To, Acc, StateName, StateData) ->
+    case is_old_session_conflict(Acc, StateData) of
+        false ->
+            process_incoming_stanza(Name, From, To, Acc, StateName, StateData);
+        true -> %% A race condition detected
+            %% Same jid, but different sids
+            SenderSID = mongoose_acc:get_prop(sender_sid, Acc),
+            ?ERROR_MSG("event=conflict_check_failed "
+                        "jid=~ts c2s_sid=~p sender_sid=~p acc=~1000p",
+                       [jid:to_binary(StateData#state.jid), StateData#state.sid,
+                        SenderSID, Acc]),
+            finish_state(ok, StateName, StateData)
+    end.
+
+%% If jid is the same, but sid is not.
+%% jid example is alice@localhost/res1.
+%% sid example is `{now(), pid()}'.
+%% The conflict can happen, when actions with an accumulator were initiated by
+%% one process but the resulting stanzas were routed to another process with
+%% the same JID but different SID.
+%% The conflict is usually happens when an user is reconnecting.
+-spec is_old_session_conflict(mongoose_acc:t(), state()) -> boolean().
+is_old_session_conflict(Acc, #state{sid = SID, jid = JID}) ->
+    SenderSID = mongoose_acc:get_prop(sender_sid, Acc),
+    SenderJID = mongoose_acc:get_prop(sender_jid, Acc),
+    (SenderJID =:= JID)
+    andalso (SenderSID =/= undefined)
```

Comment:
how about `not SenderSID == SID` instead of 2 checks?

Reply:
I've refactored the function.